Source/WebCore/page/NavigatorLoginStatus.cpp

```diff
@@ -47,40 +54,92 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
     return "NavigatorLoginStatus"_s;
 }
 
-void NavigatorLoginStatus::setLoggedIn(Navigator& navigator, Ref<DeferredPromise>&& promise)
+void NavigatorLoginStatus::setLoginStatus(Navigator& navigator, const String& status, Ref<DeferredPromise>&& promise)
 {
-    NavigatorLoginStatus::from(navigator)->setLoggedIn(WTFMove(promise));
-}
-
-void NavigatorLoginStatus::setLoggedOut(Navigator& navigator, Ref<DeferredPromise>&& promise)
-{
-    NavigatorLoginStatus::from(navigator)->setLoggedOut(WTFMove(promise));
+    NavigatorLoginStatus::from(navigator)->setLoginStatus(status, WTFMove(promise));
 }
 
 void NavigatorLoginStatus::isLoggedIn(Navigator& navigator, Ref<DeferredPromise>&& promise)
 {
     NavigatorLoginStatus::from(navigator)->isLoggedIn(WTFMove(promise));
 }
 
-void NavigatorLoginStatus::setLoggedIn(Ref<DeferredPromise>&& promise)
+bool NavigatorLoginStatus::hasSameOrigin() const
 {
-    if (m_navigator.cookieEnabled())
-        promise->resolve();
-    else
-        promise->reject();
+    RefPtr document = m_navigator.document();
+    Ref origin = document->securityOrigin();
+    bool isSameSite = true;
+    for (RefPtr parentDocument = document->parentDocument(); parentDocument; parentDocument = parentDocument->parentDocument()) {
+        if (!origin->isSameOriginAs(parentDocument->securityOrigin())) {
+            isSameSite = false;
+            break;
+        }
+    }
+    return isSameSite;
 }
 
-void NavigatorLoginStatus::setLoggedOut(Ref<DeferredPromise>&& promise)
+void NavigatorLoginStatus::setLoginStatus(const String& status, Ref<DeferredPromise>&& promise)
 {
-    promise->resolve();
+    IsLoggedIn loginStatus;
+    if (status == "logged-in"_s)
+        loginStatus = IsLoggedIn::Yes;
+    else if (status == "logged-out"_s)
+        loginStatus = IsLoggedIn::No;
+    else {
+        promise->reject();
+        return;
+    }
+
+    RefPtr document = m_navigator.document();
+    if (!document && hasSameOrigin()) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    RefPtr frame = document->frame();
+    if (!frame || !frame->isMainFrame()) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    RefPtr page = frame->page();
+    if (!page) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    page->chrome().client().setLoginStatus(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), loginStatus, [weakThis = WeakPtr { *this }, promise = WTFMove(promise)] {
+        if (!weakThis)
```

Comment:
Why do we need this check?

Reply:
I believe it ensures that the object still exists when the callback is executed. If weakThis is null, it means the object has been deleted or is no longer valid, and the callback should not proceed. 
Also I saw it's been used in `DocumentStorageAccess::requestStorageAccess` which is close to what I'm doing in terms of functionality:
```
void DocumentStorageAccess::requestStorageAccess(Ref<DeferredPromise>&& promise)
{
  ...
  page->chrome().client().requestStorageAccess(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), RegistrableDomain::uncheckedCreateFromHost(document->topOrigin().host()), *frame, m_storageAccessScope, [this, weakThis = WeakPtr { *this }, promise = WTFMove(promise)] (RequestStorageAccessResult result) mutable {
          if (!weakThis)
              return;
              ....
```

Reply:
In our case, we don't need to capture `weakThis` at all, since we are not accessing anything on `this` within the lambda, just resolving the promise.